### ISSUE TITLE
Training leads to model collapse

### ISSUE THREAD

**Issue by Ricardokevins** [princeton-nlp/SimPO#12]
Hi, Interesting work!  I immediately trained and tested it on my downstream tasks. However, the generated results showed that the model collapsed, outputting a lot of repetition, similar to "Hi Hi Hi Hi Hi". Upon checking the training log, I found that although the margin and accuracy increased, the accept reward and reject reward rapidly decreased.

To rule out the influence of my private dataset and task format, I conducted training on a small subset of the public datasets hiyouga/DPO-En-Zh-20k, and the training dynamics were similar, the reward both decreased rapidly. Have you encountered the similar issues, and how can I resolve it?

Training Argument:
```
batch 128
max_step 1000
simpo_gamma 0.5
beta 2.5
optim paged_adamw_32bit
```

Reward/Reject
![下载 (4)](https://github.com/princeton-nlp/SimPO/assets/43642508/c93f9d6c-ff4b-4299-932f-05d422387c3c)

Reward/Accept
![下载 (2)](https://github.com/princeton-nlp/SimPO/assets/43642508/cdf0d99c-b885-4e70-aeef-6c2ad5fb4bd5)

Reward/Margin
![下载 (3)](https://github.com/princeton-nlp/SimPO/assets/43642508/63ae6bc0-f8f0-44d0-8a8f-64e5d05a1529)

Reward/Accuracy
![下载 (1)](https://github.com/princeton-nlp/SimPO/assets/43642508/8cff1d70-60a0-42af-8654-928403868de0)

Training Loss
![下载 (5)](https://github.com/princeton-nlp/SimPO/assets/43642508/44a3d4cd-c7af-4c8b-959b-5d5d4e8a7d46)

**Comment by yumeng5** [comment]
Hi,

Could you share the model as well as the learning rate you used? When the model degenerates that's usually an implication of the learning rate being too high. I'd try tuning down the learning rate a bit.

Best,
Yu

**Comment by hjh0119** [comment]
I find that SimPO+LoRA performed poorly, maybe full fine-tuning is required.

**Comment by Ricardokevins** [comment]
```
# Model arguments
model_name_or_path: xxxxxxxxxx/PLM/Meta-Llama-3-8B-Instruct
torch_dtype: null
use_flash_attention_2: true

### ddp
ddp_timeout: 180000000
deepspeed: xxxxx/deepspeed/ds_z3_config.json

# Data training arguments
dataset_mixer:
  HuggingFaceH4/ultrafeedback_binarized: 1.0
dataset_splits:
- train_prefs
- test_prefs
preprocessing_num_workers: 12

# SimPOTrainer arguments
bf16: true
beta: 2.5
gamma: 1.4
do_eval: true
evaluation_strategy: steps
eval_steps: 400
gradient_accumulation_steps: 8
gradient_checkpointing: true
gradient_checkpointing_kwargs:
  use_reentrant: False
learning_rate: 1.0e-6
log_level: info
logging_steps: 5
lr_scheduler_type: cosine
max_length: 2048
max_prompt_length: 1800
num_train_epochs: 1
optim: adamw_torch
output_dir: xxxx/Trained/dev-simpo
run_name: llama-3-8b-instruct-simpo
per_device_train_batch_size: 1
per_device_eval_batch_size: 1
push_to_hub: false
save_strategy: "steps"
save_steps: 1000000
report_to:
- wandb
save_total_limit: 3
seed: 42
warmup_ratio: 0.1
report_to: wandb
run_name: simpo-repro # 可选
```



Hi, I still can not reproduce the result. Above is the training scripts

And following is the result of generate.py
```
Loading checkpoint shards: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████| 4/4 [00:02<00:00,  1.90it/s]
Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.
[{'role': 'user', 'content': "What's the difference between llamas and alpacas?"}, {'role': 'assistant', 'content': 'Both belong to camelids family.\n\nKey differences:\n\n1\\. Size: Llamas tend to grow taller, usually between five feet\n\n(\n15\n\n)&\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n–\n\n'}]
```

**Comment by Ricardokevins** [comment]
Hello, thank you very much for your reply! @yumeng5 

As I am interested in applying Simpo to my scenario, my intention is not to completely replicate the results of the paper 100%. What I am more concerned about is whether the algorithm can achieve a stable training result (temporarily regardless of the performance improvement relative to the basic SFT/Instruct models). Therefore, I am quite interested in this peculiar phenomenon where the trained model collapses directly after training. After I attempted a strict replication setup, it seems that the results are much better. However, it still seems strange that changing the dataset or some minor change would lead to training collapse. Do you have any further suggestions on this matter?

**Comment by yumeng5** [comment]
Hi @Ricardokevins 

We tested SimPO across several setups, including base setups (using off-policy preference data) and instruct setups (using on-policy preference data), with different models (llama3 and mistral). Given its stable performance across these setups, we believe that it should generalize reasonably well to other setups. Below are a few suggestions you might try:

1. Fix a large batch size (128) and tune down the learning rate. We found that the optimal learning rate varies for different datasets/setups, and using a lower learning rate usually helps prevent degeneration issues.
2. If the preference dataset is off-policy (i.e., not generated by the model being trained, as in our base setups), it might be helpful to first fine-tune the model with SFT on the preference dataset before running SimPO to reduce distribution shift.
3. Alternatively, you might add an SFT regularization loss to SimPO if the preference dataset is off-policy, which may lead to more stable training.

I hope these suggestions are helpful!

Best,
Yu

**Comment by syboomsy** [comment]
emm, I have also encountered weird model collapse phenomenon, my base model is mistral using mistral-base training config with few necessary entries modified(paths,  batch size 4gpu x 2 x 8 accumulation steps, learning rate is 3e-7),  the training log shows that the model converges well on my custom Chinese dataset (200k samples), but the inference result is a mess:
![image](https://github.com/princeton-nlp/SimPO/assets/57123221/ab28efab-3fea-4ec8-b2a7-71e8d854b3f9),
and this is training summary:
![image](https://github.com/princeton-nlp/SimPO/assets/57123221/2b97c174-95c5-465a-986b-d8898adc841d), I tried different prompts setting but it does not work, now I am trying to add SFT loss as regularization just as the way ORPO does.

**Comment by syboomsy** [comment]
> emm, I have also encountered weird model collapse phenomenon, my base model is mistral using mistral-base training config with few necessary entries modified(paths, batch size 4gpu x 2 x 8 accumulation steps, learning rate is 3e-7), the training log shows that the model converges well on my custom Chinese dataset (200k samples), but the inference result is a mess: ![image](https://private-user-images.githubusercontent.com/57123221/336744326-ab28efab-3fea-4ec8-b2a7-71e8d854b3f9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc1NzQ4MjEsIm5iZiI6MTcxNzU3NDUyMSwicGF0aCI6Ii81NzEyMzIyMS8zMzY3NDQzMjYtYWIyOGVmYWItM2ZlYS00ZWM4LWIyYTctNzFlOGQ4NTRiM2Y5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjA1VDA4MDIwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQxMTU0ZmU5ZTAzZDJkZTQ4ZWNlZTI5MTc1ZmUxM2NkZTk3NjQwZjYzYTY5NTU1Njc3Njk1YTliM2ExNjZjYzYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.wx8-gagyFLNRfxMTUPrT1jqhE6oeOkl2GMvaA6VcNn4), and this is training summary: ![image](https://private-user-images.githubusercontent.com/57123221/336744787-2b97c174-95c5-465a-986b-d8898adc841d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc1NzQ4MjEsIm5iZiI6MTcxNzU3NDUyMSwicGF0aCI6Ii81NzEyMzIyMS8zMzY3NDQ3ODctMmI5N2MxNzQtOTVjNS00NjVhLTk4NmItZDg4OThhZGM4NDFkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjA1VDA4MDIwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc5N2I4ZTczODkxNzUxNjJiYmI0Y2MxZmI4MDEzOGQ5ZGYxYjg0ZDkyNzY4MDFiMGZlOWIxNmEwMTRhNWRhMDImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.gtQV7DR51eMB89OJb48VKKxD_m9TyFiFBHZbY8tHy_k), I tried different prompts setting but it does not work, now I am trying to add SFT loss as regularization just as the way ORPO does.

extra details of my training config 
![image](https://github.com/princeton-nlp/SimPO/assets/57123221/721a05e7-b7cb-4922-878d-c403dce742a8)

**Comment by syboomsy** [comment]
It works, it seems that SimPO loss is not sufficient to ensure model stable finetuning result, adding SFT loss for balancing can solve the problem.

**Comment by zhouku92** [comment]
Hi @syboomsy 
when you said it works, do you mean SimPO + SFT loss outperforms DPO + SFT (or ORPO)? Or you simply means its performance looks reasonable.

**Comment by syboomsy** [comment]
> Hi @syboomsy when you said it works, do you mean SimPO + SFT loss outperforms DPO + SFT (or ORPO)? Or you simply means its performance looks reasonable.

SFT loss just makes performance look reasonable, and whether the SimPO loss can help model perform better needs to be evaluated.
![image](https://github.com/princeton-nlp/SimPO/assets/57123221/b3268e17-0c3f-4a09-ae98-4ac7652b80ac)
![image](https://github.com/princeton-nlp/SimPO/assets/57123221/d9fad0d9-2328-4cf7-bba6-1cdacb70473a)

**Comment by AIR-hl** [comment]
> Has anyone tried SimPO+LoRA? Does it work well?
 Hi! Have you tried SimPO + Lora? I found on AlpacaEval LC win rate, SimPO + Lora might performs better than SimPO + full. And this also happened to the DPO.
